File: 06-week-5--big-o-continued/02-day-5--implement-a-set/ruby/my_set.rb
class MySet 
  attr_reader :data
  # throw an error if called with anything other than string, array or nothing
  # if an iterable is provided only its unique values should be in data
  # strings and arrays will need to be broken down by their elements/characters
  def initialize(iterable = nil)
    @data = {}
  end

  # return number of elements in MySet
  def size
  end

  # add an item to MySet as is
  # return the MySet instance
  def add(item)
  end

  # delete an item from MySet
  # return true if successful otherwise false
  def delete(item)
  end

  # return true if in MySet, otherwise false
  def has(item)
  end

  # return data as an array
  def entries
  end
end

if __FILE__ == $PROGRAM_NAME
  # Don't forget to add your own!
end

File: 06-week-5--big-o-continued/02-day-5--implement-a-set/javascript/my_set.js
class MySet {
  // throw an error if called with anything other than string, array or nothing
  // if an iterable is provided only its unique values should be in data
  // strings and arrays will need to be broken down by their elements/characters
  constructor(iterable) {
    if (!(iterable === undefined || 
      Array.isArray(iterable) ||
      typeof iterable === "string")) {
        throw new Error("MySet only accepts iterable or nothing on initialization")
      } 



    this.data = {};

    if (iterable) {
      for (const e of iterable) {
        this.data[e] = true
      }
    }



  }

  // return number of elements in MySet
  size() {
    return this.entries().length
  }

  // add an item to MySet as is
  // don't worry about arrays here!
  // return the MySet instance
  add(item) {
    this.data[item] = true
    return this
  }

  // delete an item from MySet
  // don't worry about arrays here!
  // return true if successful, otherwise false
  delete(item) {
    if (this.has(item)) {
      delete this.data[item]
      return true
    } else {
      return false
    }

  }
  // return true if in MySet, otherwise false
  // don't worry about arrays here!
  has(item) {
   return !!this.data[item]
  }

  // return data as an array
  // don't worry about arrays here!
  entries() {
    return Object.keys(this.data)
  }
}

if (require.main === module) {
  // add your own tests in here
  
}

module.exports = MySet;

